Pangolin/Framework/Messaging/MessageQueue.cs

```diff
@@ -0,0 +1,199 @@
+﻿using System;
+using System.Data.SqlClient;
+using EnderPi.Framework.Caching;
+using System.Text.RegularExpressions;
+
+namespace EnderPi.Framework.Messaging
+{
+    /// <summary>
+    /// This is a message queue backed by SQL server.  Simple API.
+    /// </summary>
+    public class MessageQueue : IMessageQueue
+    {        
+        private static readonly Regex regex = new Regex("^[a-zA-Z0-9]*$");
+
+        private string _queueName;
+        private string _databaseConnection;
+        private const string _createTableStatement = @"CREATE TABLE MessageQueue.{0} (Id BIGINT IDENTITY(1,1), Priority int, DateCreated DateTime , MessageBody VARCHAR(MAX)) CREATE CLUSTERED INDEX [MessageIndex] ON MessageQueue.{0} ([Priority] DESC, [ID] DESC) WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, SORT_IN_TEMPDB = OFF, DROP_EXISTING = OFF, ONLINE = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF)";
+        private const string _tableExistsStatement = @"SELECT OBJECT_ID('MessageQueue.{0}')";
+        private const string _insertMessageStatement = @"INSERT INTO [MessageQueue].{0} ([Priority],[DateCreated],[MessageBody]) VALUES (@Priority,@DateCreated,@MessageBody)";
+        private const string _readMessageStatement = @"WITH T AS (SELECT TOP 1 [Id],[Priority],[DateCreated],[MessageBody] FROM MessageQueue.{0} ORDER BY PRIORITY DESC, ID DESC) DELETE FROM T OUTPUT DELETED.ID, DELETED.Priority, DELETED.DateCreated, DELETED.MessageBody";        
+        private const string _peekMessageStatement = "SELECT TOP 1 [Id],[Priority],[DateCreated],[MessageBody] FROM MessageQueue.{0} ORDER BY PRIORITY DESC, ID DESC";
+        private const string _countMessagesStatement = "SELECT COUNT(*) FROM MessageQueue.{0}";
+        private const string _clearMessagesStatement = "DELETE FROM MessageQueue.{0}";
+
+        public string Name => _queueName;
+
+        /// <summary>
+        /// Constructs the object, builds the underlying table if it doesn't exist.
+        /// </summary>
+        /// <param name="Name">The name of the queue</param>
+        public MessageQueue(string databaseConnection, string queueName, ICache cache)
+        {
+            if (!IsQueueNameValid(queueName))
+            {
+                throw new ArgumentException("Queue name is not valid");
+            }
+            _databaseConnection = databaseConnection;
+            _queueName = queueName;
+            //TODO I don't know that this even needs to cache.  Users shouldn't be creating message queues frequently.  They will likely pass one around as a dependency.
+            if (!TableExists(cache))
```

Comment:
Any thought about maybe making a maximum ceiling to prevent creating too many tables or message queues?

Reply:
This is a good idea.  I want it to be configurable, so I"m not going to do it now, but I will in the next couple of cards, as those involve adding configuration key-value pair tables.  I'll make the change then, and default it to like 128 or 256 or something.